# Question 1

In [ ]:
""" 
It's formula is;
  (input_shape+1) * 512
  +
  513 * 512
  +
  513 * 512
  +
  513 * (output_shape), 
  if I'm not mistaken.
"""

# In general, we add +1 to the former layer in the multiplication step between two layers.


In the lectures we've seen that how to get the number of parameters so I'll use them to calculate first question's answer with a model.

In [ ]:
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [ ]:
from tensorflow.keras.models import Sequential

model = Sequential()


In [ ]:
from tensorflow.keras.datasets import boston_housing

(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

print("Shape of train set (X) :", X_train.shape)
print("Shape of train set (y) :", y_train.shape)
print("Shape of test set  (X) :", X_test.shape)
print("Shape of test set  (y) :", y_test.shape)

65536/57026 [==================================] - 0s 0us/step
Shape of train set (X) : (404, 13)
Shape of train set (y) : (404,)
Shape of test set  (X) : (102, 13)
Shape of test set  (y) : (102,)


In [ ]:
from tensorflow.keras.layers import Dense

input_number = X_train.shape[1] 

model = Sequential()

model.add(Dense(512, activation='relu', input_shape=(input_number,), name = "Hidden_Layer_1"))
model.add(Dense(512, activation='relu', name = "Hidden_Layer_2"))
model.add(Dense(512, activation='relu', name = "Hidden_Layer_3"))
model.add(Dense(1, name = "Output"))

model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])

model.fit(X_train, y_train, epochs=20, verbose=0)

In [ ]:
input_number

13

In [ ]:
mae_train = model.evaluate(X_train, y_train, verbose = 0)[1]
mae_test = model.evaluate(X_test, y_test, verbose = 0)[1]

print("\nMAE for train set : {:.3f}".format(mae_train))
print("MAE for test set    : {:.3f}".format(mae_test))


MAE for train set : 4.780
MAE for test set    : 5.336


In [ ]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 Hidden_Layer_1 (Dense)      (None, 512)               7168      
                                                                 
 Hidden_Layer_2 (Dense)      (None, 512)               262656    
                                                                 
 Hidden_Layer_3 (Dense)      (None, 512)               262656    
                                                                 
 Output (Dense)              (None, 1)                 513       
                                                                 
Total params: 532,993
Trainable params: 532,993
Non-trainable params: 0
_________________________________________________________________


Hence, for 3 layers ANN with 412 neurons in each hidden layer, we have 532,993 parameters.

# Question 2

I guess, for 2 input variables, if we use modulo 2 for each column, we can get the function XOR in a single layer. It'll be implemented as;


*   If, sum of a column = 0 (mod2):
      then XOR returns 1
*   Else:
      XOR returns 0.



# Question 3

In [ ]:
# y = -17 in that RELU question